portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java

```diff
@@ -30,29 +35,58 @@
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // If user is logged in and has not set a nickname, redirect to nickname page
     UserService userService = UserServiceFactory.getUserService();
+    if (userService.isUserLoggedIn()) {
+      
+    }
+    
+    // Create variables needed to create a user.
     Boolean loggedIn = false;
     String authenticationUrl = ""; // URL used for either login or logout.
     String userEmail = "";
+    String nickname = "";
     String redirectUrl = "/"; // Both login and logout redirect to the same URL
 
+    // Define values for user variables depending on login status.
     if (userService.isUserLoggedIn()) {
-      // Create a new user with logout URL and email.
       loggedIn = true;
-      authenticationUrl = userService.createLogoutURL(redirectUrl);
       userEmail = userService.getCurrentUser().getEmail();
+      nickname = getUserNickname(userService.getCurrentUser().getUserId());
+      if (nickname == null) {
+        // If logged in user has no nickname, redirect to nickname setup page.
+        nickname = "";
+        authenticationUrl = "/nickname.html";
+      } else {
+        // If logged in user has nickname, set login URL.
+        authenticationUrl = userService.createLogoutURL(redirectUrl);
+      }
     } else {
-      // Create a new user with login URL and no email.
       authenticationUrl = userService.createLoginURL(redirectUrl);
     }
 
     // Create new user.
     UserAuthenticationData userAuthenticationData =
-        new UserAuthenticationData(loggedIn, authenticationUrl, userEmail);
+        new UserAuthenticationData(loggedIn, authenticationUrl, userEmail, nickname);
 
     // Convert the user to JSON format and return it as response.
     Gson gson = new Gson();
     response.setContentType("application/json");
     response.getWriter().println(gson.toJson(userAuthenticationData));
   }
+
+  /** Returns the nickname of the user with id, or null if the user has not set a nickname. */
+  private String getUserNickname(String id) {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Query query =
+        new Query("UserInfo")
+            .setFilter(new Query.FilterPredicate("id", Query.FilterOperator.EQUAL, id));
+    PreparedQuery results = datastore.prepare(query);
+    Entity entity = results.asSingleEntity();
+    if (entity == null) {
+      return null;
+    }
+    String nickname = (String) entity.getProperty("nickname");
+    return nickname;
+  }
```

Comment:
Typically, it's not ideal to have duplicate code in multiple places (here, NicknameServlet and DataServlet.java). For example, imagine if a future developer updated this method but forgot to update it in DataServlet.java and NicknameServlet - there could be mismatched logic and unfortunate bugs :(

Rather, you could declare this as a shared public utils method that both servlets can access. Then each servlet that needs the function can import that utils class and use the same function. If you're not sure how to go about this let me know!

Reply:
Done! Ended up implemented the getNickname() function from the Users API!